packages/mui-base/src/useNumberInput/useNumberInput.ts

```diff
@@ -0,0 +1,371 @@
+        event.stopPropagation();
+        return;
+      }
+
+      otherHandlers.onFocus?.(event);
+
+      if (event.defaultPrevented) {
+        return;
+      }
+
+      if (formControlContext && formControlContext.onFocus) {
+        formControlContext?.onFocus?.();
+      }
+      setFocused(true);
+    };
+
+  const handleValueChange =
+    () =>
+    (
+      event: React.FocusEvent<HTMLInputElement> | React.PointerEvent | React.KeyboardEvent,
+      val: number | undefined,
+    ) => {
+      // 1. clamp the number
+      // 2. setInputValue(clamped_value)
+      // 3. call onValueChange(newValue)
+      let newValue;
+
+      if (val === undefined) {
+        newValue = val;
+        setInputValue('');
+      } else {
+        newValue = clamp(val, min, max, step);
+        setInputValue(String(newValue));
+      }
+
+      setValue(newValue);
+
+      // TODO: integration with formControlContext
+      //       OR: (event, newValue) similar to SelectUnstyled
+      // formControlContext?.onValueChange?.(newValue);
+
+      if (isNumber(newValue)) {
+        onValueChange?.(newValue);
+      } else {
+        onValueChange?.(undefined);
+      }
+    };
+
+  const handleInputChange =
+    (otherHandlers: Record<string, React.EventHandler<any> | undefined>) =>
+    (event: React.ChangeEvent<HTMLInputElement>) => {
+      if (!isControlled && event.target === null) {
+        throw new MuiError(
+          'MUI: Expected valid input target. ' +
+            'Did you use a custom `slots.input` and forget to forward refs? ' +
+            'See https://mui.com/r/input-component-ref-interface for more info.',
+        );
+      }
+
+      formControlContext?.onChange?.(event);
+
+      otherHandlers.onChange?.(event);
+
+      const val = parseInput(event.currentTarget.value);
+
+      if (val === '' || val === '-') {
+        setInputValue(val);
+        setValue(undefined);
+      }
+
+      if (val.match(/^-?\d+?$/)) {
+        setInputValue(val);
+        setValue(parseInt(val, 10));
+      }
+    };
+
+  const handleBlur =
+    (otherHandlers: Record<string, React.EventHandler<any> | undefined>) =>
+    (event: React.FocusEvent<HTMLInputElement>) => {
+      const val = parseInput(event.currentTarget.value);
+
+      otherHandlers.onBlur?.(event);
+
+      if (val === '' || val === '-') {
+        handleValueChange()(event, undefined);
+      } else {
+        handleValueChange()(event, parseInt(val, 10));
+      }
+
+      if (formControlContext && formControlContext.onBlur) {
+        formControlContext.onBlur();
+      }
+
+      setFocused(false);
+    };
+
+  const handleClick =
+    (otherHandlers: Record<string, React.EventHandler<any>>) =>
+    (event: React.MouseEvent<HTMLInputElement>) => {
+      if (inputRef.current && event.currentTarget === event.target) {
+        inputRef.current.focus();
+      }
+
+      otherHandlers.onClick?.(event);
+    };
+
+  const handleStep =
```

Comment:
Just an idea to consider: this looks like a good use case for a [reducer](https://github.com/mui/material-ui/blob/75c1dfb60febe05acf4e4e4c7963426f101f31b0/packages/mui-base/src/utils/useControllableReducer.ts)

Reply:
I started trying this out but I think its better to finish it in a separate PR ~